pkg/antctl/raw/featuregates/command.go

```diff
@@ -0,0 +1,180 @@
+package featuregates
```

Comment:
Sorry, I might have missed something, but why it needs to be a "raw" command, if it just query controller or agent API to get the feature gate information?

Reply:
I was wondering the exact same thing. My understanding is that if we want *not* to make it a raw command, we need it to be an aggregated API (if we want the out-of-cluster / remote mode to work), which in turns means it needs to be in an API group for which we have an APIService (such as `system`). If we don't want to introduce a new versioned API for this or if we want to avoid relying on aggregation, it seems that it does need to be a "raw" command.
cc @tnqn

Reply:
yes, I confirmed with @tnqn  last time, he doesn't suggest to use aggregated API which will introduces a lots of auto-generated codes and it's too heavy for a simple command. and it need to be a 'raw' command if we didn't involve an aggregated API.

Reply:
But if it is not aggregated, then antctl cannot call the API through K8s API IP, but it must be able to reach the controller Node IP?
@tnqn : generally I wonder what is your opinion what API/commands should go through API aggregation. And by "heavy" what is the major concern - binary size, memory overhead?

Reply:
If we aggregate an API, here is what it requires or affects I can think of:
1. A versioned API and all related source files: types.go, doc.go, register.go, deepcopy.go
2. Clientset code, OpenAPI code generated from it (8 files in total).
3. "featuregate" will be listed in `kubectl api-resources` as a resource.

In general, I think API should go to its category literally, i.e. according to whether it's a "resource", unless there is a strong reason to make an exception, like it's strongly required to be accessed via K8s API or it requires the versioned API machinery to keep backwards-compablity. IMO Feature gate is just a configuration nothing different from loglevel and version API. I'm not sure what's the strong reason it needs to be accessed via K8s API separately and it would work for controller only anyway. Why don't we document an unified way to get such information of an antrea component via antctl inside the Pods, instead of listing several scenarios: "if you want to get runtime feature gate configuration of antrea-controller, you can execute antctl outside the cluster...; if you want the same of antrea-agent, you need to execute antctl inside the antrea-agent pod, e.g."

Maybe "heavy" didn't express my opinion accurrently, I meant "unworthy and unnaturally".

And for current way to get it from outside-of-cluster, how does it "print both Controller and Agent's feature gates list"? Is there still an issue that the agent's configuration can only be read from configmap and the actual enablement may be wrong because of the version skew?

Reply:
Ok, I found I might have misunderstanding before. I thought all controller commands can be executed out-of-Pod, but log-level and query can not be executed outside either? If we have no good way to expose all controller APIs through K8s apiserver, then I agree no need to support that for featuregates either. But then do we still want to get featuregates out-of-Pod using the Node IP? If yes, then probably we should implement some common code for reaching controller/agent API through that way, rather than adding such code for individual commands? @tnqn @antoninbas @luolanzone

Reply:
@jianjuns I have moved a few shared functions into a helper.go before, so far I saw featuregate and supportbunddle will reach controller/agent API directly, but their behaviors are different, so I didn't merge then, I will check if there are more codes we can reuse between different raw commands.

Reply:
Thanks @luolanzone for the update. @antoninbas @tnqn : do you think we should handle remote access separately if we believe we should support that for featuregates, or even more commands (I actually feel if we want to enable it for one command, then why not build a common way to enable it for all)?

Reply:
@luolanzone is doing some refactoring in this PR to re-use some code, so that works for me
I agree that having some framework that makes writing such commands easier would be beneficial, but I don't think it needs to be in this PR. I am also not sure what's the best approach long-term: we can continue to connect to the Node directly (which assumes that antctl is run on a machine that has connectivity to the Node and that the API port used by Antrea is not blocked by a firewall), or we could investigate something else, such as going through the K8s API with the exec verb for the antrea-controller Pod (basically "kubectl exec"). One more issue with the current approach is that the connection to the controller is insecure and we would need to retrieve the antrea-ca from the K8s API to change that, which is possible but means added complexity compared to using "exec".
BTW, I do think that there is value in being able to run antctl out-of-cluster for several commands (but maybe not all), as opposed to requiring users to run "kubectl exec" themselves. It also gives us the opportunity to differentiate between out-of-cluster and in-Pod behavior in the implementation if we ever need to. But maybe our implementation should be simplified and use "exec" itself.

Reply:
I feel if we want to support accessing controller API through Node IP, adding a command for that should be as simple as other commands (those through API aggregation or those support only in-Pod), so no need to add raw command code at all, and yes we should handle authentication too. So we have three choices for this PR: 1) do not support out-of-cluster; 2) merge the current code which implements out-of-cluster but with raw command; 3) implement a common framework for out-of-cluster via Node IP and remove raw command code.

I would prefer 1), but I am ok with 2) too.

Reply:
> I feel if we want to support accessing controller API through Node IP, adding a command for that should be as simple as other commands

I agree.

I would prefer solution 2) since the code is already there (it was needed for other commands) and I feel like having the ability to run this command out-of-cluster is good. With a follow-up PR to define the framework for out-of-cluster access and refactor existing commands.

Reply:
@antoninbas @jianjuns @tnqn , do we have any conclusion for this feature? I suppose we can raise a new issue for a common out-of-cluster access framework?

Reply:
I am fine with merging the current way too.

Reply:
@antoninbas @jianjuns @tnqn  Could you help to double check this PR so we can move forward? thanks!

Reply:
please add licenseheader